io/epoll.cpp

```diff
@@ -273,35 +265,39 @@ class EventEngineEPoll : public MasterEventEngine, public CascadingEventEngine,
     }
     virtual int cancel_wait() override { return eventfd_write(_evfd, 1); }
 
-    int wait_for_fd(int fd, uint32_t interests, uint64_t timeout) override {
-        Event event{fd, interests | ONE_SHOT, CURRENT};
-        int ret = add_interest(event);
+    int wait_for_fd(int fd, uint32_t interest, uint64_t timeout) override {
+        if (fd < 0)
+            LOG_ERROR_RETURN(EINVAL, -1, "invalid fd");
+        if (interest & (interest-1))
+            LOG_ERROR_RETURN(EINVAL, -1, "can not wait for multiple interests");
+        if (unlikely(!interest))
+            return rm_interest({fd, EVENT_RWE, 0}); // remove fd from epoll
+        int ret = add_interest({fd, interest | ONE_SHOT, CURRENT});
         if (ret < 0) LOG_ERROR_RETURN(0, -1, "failed to add event interest");
         ret = thread_usleep(timeout);
         ERRNO err;
         if (ret == -1 && err.no == EOK) {
             return 0;  // Event arrived
-        } else if (ret == 0) {
-            rm_interest(event);  // Timeout
-            errno = ETIMEDOUT;
-            return -1;
-        } else {
-            rm_interest(event);  // Interrupted by other thread
-            errno = err.no;
-            return -1;
         }
```

Comment:
I think the previous if...else is more self-evident.

People would clearly see that 0 is timeout, -1 is interrupted, etc.  More educational.